File: python/day06/day6.py
import pathlib
import re

CWD = pathlib.Path(__file__).parent.absolute()
filename = pathlib.PurePath(CWD, 'data')
with open(filename) as f:
    data = [x.strip() for x in f.read().split('\n\n')]

inters = []
unions = []
for group in data:
    groupsets = [set(ch for ch in person) for person in group.splitlines()]
    inters.append(set.intersection(*groupsets))
    unions.append(set.union(*groupsets))

print(f'Part 1: {sum(len(x) for x in unions)}')
print(f'Part 2: {sum(len(x) for x in inters)}')

File: js/day06/day6.js
import fs from 'fs'
import path from 'path'

let filename = path.join(path.dirname(process.argv[1]), 'data')
let data = fs.readFileSync(filename, 'utf8')
    .trim()
    .split('\n\n')
    .filter(x => x !== '')

let inters = []
let unions = []
for (let group of data) {
    let groupsets = group.split('\n').map(person => new Set([...person]))
    unions.push(groupsets.reduce((x, y) => new Set([...x, ...y])))
    inters.push(groupsets.reduce((x, y) => new Set([...x].filter(a => y.has(a)))))
}

console.log(unions.map(x => x.size).reduce((x, y) => x+y))
console.log(inters.map(x => x.size).reduce((x, y) => x+y))
